dali/daliadmin/daadmin.cpp

```diff
@@ -2264,6 +2264,98 @@ void getxref(const char *dst)
     conn->close();
 }
 
+void checkFileSizeOne(IUserDescriptor *user, const char *lfn, bool fix)
+{
+    try
+    {
+        CDfsLogicalFileName dlfn;
+        dlfn.set(lfn);
+        Owned<IDistributedFile> dFile = queryDistributedFileDirectory().lookup(dlfn, user, AccessMode::tbdRead, false, false, nullptr, defaultPrivilegedUser, 30000); // 30 sec timeout
+        if (dFile)
+        {
+            bool locked = false;
+            COnScopeExit ensureUnlock([&]() { if (locked) dFile->unlockProperties(); });
+            if (dFile->querySuperFile())
+                WARNLOG("Skipping: file '%s' is a superfile", lfn);
+            else
+            {
+                unsigned numParts = dFile->numParts();
+                for (unsigned p=0; p<numParts; p++)
+                {
+                    IDistributedFilePart &part = dFile->queryPart(p);
+                    IPropertyTree &attrs = part.queryAttributes();
+                    if (!attrs.hasProp("@size"))
+                    {
+                        if (fix)
+                        {
+                            offset_t partSize;
+                            try
+                            {
+                                partSize = part.getFileSize(true, true);
+                            }
+                            catch (IException *e)
+                            {
+                                EXCLOG(e);
+                                e->Release();
+                                continue;
+                            }
+                            if (!locked)
+                            {
+                                part.lockProperties(30000);
+                                locked = true;
```

Comment:
This is using knowledge of the internal implementation to avoid locking multiple times.  Should it lock the file instead to make it explicit??
also, there should be a corresponding unlockProperties.

Reply:
I agree, locking just the file isn't really right either - that would also be using knowledge of internal implementation (i.e. that locking file equates to part attributes being protected too).
I will modify.

>also, there should be a corresponding unlockProperties.

there is, via COnScopeExit ensureUnlock.

Reply:
>I agree, locking just the file isn't really right either - that would also be using knowledge of internal implementation (i.e. that locking file equates to part attributes being protected too).
I will modify.

it was also wrong (but harmless) in the sense that the underlying implementation relies on knowing if each part is 'dirty', so should lock each part as it goes (I've changed it). It's a NOP though, if the file is already locked (which I've changed it to be).